### DISCUSSION TITLE
In-template variable definition

### DISCUSSION THREAD

**Discussion by wspl** [vuejs/rfcs#505] [General]
Sometimes you will encounter such a situation:
```vue
<template>
  <div class="container">
    <p class="row">{{item && item.a && item.a.b && item.a.b.hello}}</p>
    <p class="row">{{item && item.a && item.a.b && item.a.b.world}}</p>
    <p class="row">{{item && item.a && item.a.b && item.a.b.goodbye}}</p>
  </div>
</template>
```

These code with the same logic need to be written many times.
I hope there is a way to define local variables in the template, then the above code can be abbreviated as:

```vue
<template>
  <div class="container">
    <var name="b" :value="item && item.a && item.a.b"></var>
    <p class="row">{{b && b.hello}}</p>
    <p class="row">{{b && b.world}}</p>
    <p class="row">{{b && b.goodbye}}</p>
  </div>
</template>
```

or

```vue
<template>
  <div class="container" v-var:b="item && item.a && item.a.b">
    <p class="row">{{b && b.hello}}</p>
    <p class="row">{{b && b.world}}</p>
    <p class="row">{{b && b.goodbye}}</p>
  </div>
</template>
```

Q: Why not computed?
A: The `item` in the above code may come from `v-for`, so `computed` is not a feasible solution.

**Comment by posva** [comment]
It was proposed in the past: https://github.com/vuejs/vue/pull/7325
Currently you can achieve it with https://github.com/posva/vue-local-scope

**Comment by skyrpex** [comment]
Probably you already know, but if you use JSX you could do this:

```jsx
export default {
    render() {
        const b = this.item && this.item.a && this.item.a.b;
        return (
            <div class="container">
                <p class="row">{b && b.hello}</p>
                <p class="row">{b && b.world}</p>
                <p class="row">{b && b.goodbye}</p>
            </div>
        )
    }
}
```

**Comment by afontcu** [comment]
> Q: Why not computed?
> A: The `item` in the above code may come from `v-for`, so `computed` is not a feasible solution.

You can also create a method that accepts a parameter, so you can use it in `v-for` loops.

**Comment by phiter** [comment]
The issue with methods is that they aren't cached, so a bit less performatic.

**Comment by LinusBorg** [comment]
~Creating a local variable as proposed by OP would also not be cached.~

I see, and a  local variable could serve as an in-loop cache for method results where computed props can't be used.

**Comment by ReinisV** [comment]
> Q: Why not computed?
A: **The item in the above code may come from v-for**, so computed is not a feasible solution.

Whenever I hear someone say that, I think "this is a perfect situation for creating a component for that item!". That would give you the ability to use simple computed getters that can be cached. 

A component should always have exactly one reason to exist and one mission to execute. This parent components mission is not knowing how to render the child items.

**Comment by LinusBorg** [comment]
Well, component instantiations come at a price (memory-footprint & render performance), so sometimes it's better to render i list's items in the parent for that reason. There shuold be ways to keep these situations performant, an this RFC aims to address it.

**Comment by frehner** [comment]
Forgive my ignorance, I'm new to the vue community. 

Could the above be done with optional chaining as well? e.g. 

```html
<template>
  <div class="container">
    <p class="row">{{ item?.a?.b?.hello }}</p>
    <p class="row">{{ item?.a?.b?.world }}</p>
    <p class="row">{{ item?.a?.b?.goodbye }}</p>
  </div>
</template>
```

it's [stage 3](https://github.com/tc39/proposal-optional-chaining) and has a [babel plugin](https://babeljs.io/docs/en/babel-plugin-proposal-optional-chaining)

**Comment by LinusBorg** [comment]
Not right now, as vue-es2015-template compiler doesn't aupport >es6 features

But that can be solved.

**Comment by michaeldrotar** [comment]
I was dealing with a similar concern recently in my work so wanted to add 2 cents...

(My day job is in Ember so I'll pseudo-code the concepts)

Generally I want to pass a data model to my components.. like pass a user to a component for showing the user's badge or pass a meeting to a component for showing when and where and who's attending that meeting... it's just simpler to do `<my-component meeting=meeting>` than to do `<my-component name=meeting.name date=meeting.date whatever=meeting.whatever ...>`

I've also recently come to appreciate that this allows me to flatten out deeply nested data and to add logic around data that's not directly suited to my use case. So I recently did something like this:

```js
// Component receives `model`
name = model.model_name
date = computed(() => (model.date && date_from_format(model.date, 'YYYY-MM-DD')) || undefined)
instructions = computed(() => (model.location && model.location.meeting_instructions) || undefined)
```

The same can be done for lists... if I need to display a list of `model.users`, I originally considered passing them to a nested component as well, but sometimes that's overkill.

```js
users = computed(() => {
  return model.users.map((user) => {
    const spouse = model.spouse || {};
    return {
      name: user.name,
      spouseName: spouse.name
    };
  });
});
```

Now the template can iterate `users` without fear of accessing a nested property that might not exist and my component js continues to uphold the contract that it's explicitly handling all the data that the template needs.

**Comment by cawa-93** [comment]
Will this make the code much less clear? In my opinion, if you have a large component, it will be difficult to understand what the variable is, where it comes from if it is not in the `data` of component.

Much simpler code that doesn't have any "magic" variables:
```vue
<template>
  <div class="container" v-if="item && item.a && item.a.b">
    <p class="row">{{item.a.b.hello}}</p>
    <p class="row">{{item.a.b.world}}</p>
    <p class="row">{{item.a.b.goodbye}}</p>
  </div>
</template>
```

**Comment by woothu** [comment]
Computed can also solve v-for performance problem, as items could be mapped to have additional keys for display purposes + you can use object destructuring. 

Btw. most of times proper Item structurizing can solve this issue.

**Comment by chriswoodle** [comment]
I think I found some kind of way to accomplish in-template variables in vue 3. Here is an example where I create the "in-template variable" inside of a v-for. Basically declare an inline array and iterate over the single item.

```html
<div v-for='item in items'>
    <div v-for="record of [cache.find(item.otherId).value]">
        <span v-if="record">{{ record.name }}</span>
    </div>
</div>
```

In this case `cache.find(item.otherId)` returns back a `ref<{ name: string } | null>`

I feel like it's its kinda hacky, and I'm not really sure the performance impact.

**Comment by julie777** [comment]
> 
> 
> The issue with methods is that they aren't cached, so a bit less performatic.

Which is why there are computed functions that do cache results. :-)

**Comment by liho00** [comment]
`<div v-data="{x:'my temp variable'}">{{x}}</div>`

**Comment by komonjlep** [comment]
> I think I found some kind of way to accomplish in-template variables in vue 3. Here is an example where I create the "in-template variable" inside of a v-for. Basically declare an inline array and iterate over the single item.
> 
> ```
> <div v-for='item in items'>
>     <div v-for="record of [cache.find(item.otherId).value]">
>         <span v-if="record">{{ record.name }}</span>
>     </div>
> </div>
> ```
> 
> In this case `cache.find(item.otherId)` returns back a `ref<{ name: string } | null>`
> 
> I feel like it's its kinda hacky, and I'm not really sure the performance impact.

It's great how this approach can be used to cache cross-referenced (calculated) item when I have two v-for for matching correct item from big list. This way I don't need to re-evaluate fetching of correct item for each action separately, but can simply fetch item only once and refer to it.

It looks like this works correctly, even with events. I would hope to see this implement in more readable manner and without using v-for in hackish-way. Please consider implementing this as built-in-feature.

**Comment by dima-hx** [comment]
There is also a way to set multiple variables in `:set` directive in such way:


                            <div :set="br = {variable1: someCalculation(), variable2: anotherCalculations()}">
                                <pre>{{br}}</pre>
                            </div>

https://stackoverflow.com/a/56249209/2114398

https://dev.to/pbastowski/comment/7fc9

**Comment by myWsq** [comment]
Svelte implements this feature and has good Typescript support. 

https://svelte.dev/docs#template-syntax-const

**Comment by mon-jai** [comment]
> `<div v-data="{x:'my temp variable'}">{{x}}</div>`

I can't find any documentation on `v-data`.

**Comment by darrylnoakes** [comment]
> I can't find any documentation on `v-data`.

It was just a suggestion for the syntax, I believe.

**Comment by DevilTea** [comment]
I've a solution that is using a component to solve this issue.

> Only tested in vue >= 3.3.0

https://www.npmjs.com/package/vue-temp-var

Example
```vue
<script setup lang="ts">
import TempVar from 'vue-temp-var'

function getRandomNumber() {
  return Math.random()
}
</script>

<template>
  <div v-for="i in 10" :key="`random-${i}`">
    <!-- cause the return value would be random, we should store it somewhere -->
    <!-- bind a prop 'define' to 'TempVar'. Because it is a prop, it would still be reactive -->    
    <!-- destruct from slot: defined, the type would be kept -->
    <TempVar 
      :define="{ randomNum: getRandomNumber() }"
      #defined="{ randomNum }"
    >
      <span>"{{ randomNum }}"</span>
      <span v-if="randomNum > 0.5"> is larger than 0.5</span>
      <span v-else> is smaller than 0.5</span>
    </TempVar>
  </div>
</template>
```

**Comment by unrevised6419** [comment]
**LocalScope.vue**
```vue
<template>
	<slot v-bind="$attrs" />
</template>
```

```vue
<template>
	<LocalScope v-slot="{ classes }" classes="py-3 font-semibold">
		<p :class="classes">some text</p>
		<p :class="classes">some text</p>
	</LocalScope>
</template>
```

**Reply by DevilTea** [reply]
This looks more neat! It's not necessary to use another template to consume bounded slot props! 

But unfortunately, this method would not keep the type of the local scope variables.🤔

**Reply by unrevised6419** [reply]
I suppose this now can be achieved using generics component 🤔 

https://vuejs.org/api/sfc-script-setup.html#generics

**Reply by DevilTea** [reply]
Yes, my solution is a component that is achieved by using generics component. I just wrapped it and distributed it to npm 😂.

Here is the component's source code
```vue
<script setup lang="ts" generic="T extends Record<any, any>">
defineProps<{ define: T }>()
</script>

<template>
	<slot v-bind="(define as T)" />
</template>
```

**Reply by unrevised6419** [reply]
Strange why `defineProps<T>()` cannot be used 🤔 it does not work

```
Error: [@vue/compiler-sfc] Unresolvable type reference or unsupported built-in utility type

src/Comp.vue
4  |  
5  |  <script setup lang="ts" generic="T">
6  |  defineProps<T>()
   |              ^
7  |  </script>
```

**Reply by unrevised6419** [reply]
I found out how to avoid that error and created a lib for this component.

https://www.npmjs.com/package/@allindevelopers/vue-local-scope

**Comment by dfidalg0** [comment]
I guess the most simple solution looks like this

```vue
<!-- local-scope.vue -->
<script setup lang="ts" generic="T extends Record<string, unknown>">
defineProps<{ set: T }>();
</script>

<template>
  <slot v-bind="set" />
</template>
```

And you can use it like this

```vue
<!-- my-component.vue -->
<script setup lang="ts">
import LocalScope from './local-scope.vue';
</script>

<template>
  <local-scope
    :set="{ someVar: someValue }"
    v-slot="{ someVar }"
  >
    {{ someVar }}
  </local-scope>
</template>
```

The thing is - it's overhead having a component lifecycle just to define a variable while in JSX you could easilly define a variable inside a function.

Maybe we could have something like `v-def="{ someVar: someValue }"` that transpiles to `const someVar = someValue` inside the render function.

Or, to make it look more like standard JS, we could have

`v-def="someVar = someValue, anotherVar = anotherValue"`

transpiled to

`const someVar = someValue, anotherVar = anotherValue`

And, with Typescript,

```
v-def="
  someVar: number = someValue,
  anotherVar = anotherValue as string
"
```

would just work.

**Comment by fardolieri** [comment]
I use a functional component and force (sort of roughly) a type onto it. That way, consumers can choose a fitting name without having to rename the variable. It even allows multiple assignments. What do you think?

[See the demo.](https://play.vuejs.org/#eNqVVE1v2zAM/SucMKBJ4Thb05OXZuiGAuuwL2zFLvMOqiMnamzRkGgnReD/Pkq2267b2vVimHyk+B5Fai9OqypuaiUSMXeZ1RWBU1RXi9ToskJLcKHK6ru0kFss4SCe9vbBq9TMp10KB7NBDBSSFFsA8/Vs8U422qxAgkfQSnsNjbRaXhYKtIHyOgA+BbSDlVWSns2nnBgOGOomOeJJKj5KWsdWmiWWo3EqoJm4AomBPXAAtKkIaZy41M1i33vb+dSbDyPzQVPQ0XE/J8ikAVk4BJKbjnBdkK6Y/KDCxf9NN7mU9lEVEXDUk6SE+EekfECsXKdki3bzJ+VmkqMnZwBz+PEigpcRHEUw++l5q13F0BEcHoLx9hJrVh6iD+HodwkdFgHn3FdhEmZrmGsf5M3uz/sq3Co7wXxCW/RIOOAfsubTO4PGpqPrwv+uZ7D3saW0K20ml0iEZQLxsSp5VNswrSFURIJchibXq/jKoeHRD4mpyLCsdKHs54o0GpeKpDvSY7IocPs++MjWKhr82Vplm7/4r9zO+1LxxSqnbKNScYMRU1TUwWffPqkd/9+AJS7rgqMfAL8qh0XtOXZhb2qzZNp34gLb87DAvIIX7mxHyrhBlCfqI9sQnwre/7cPSL+lO4uPQx73k7vY30lMjlvYvxZ8qyrXJpyHRhmCtn85uIh/M1LDrXe3z8rJ/YSRrzCqLA9tBBntxnCygFH4shX7YXMx50jextfxyPskkXXjIGbPi7pWVtOp9yWQ89zziDE2BumgNhuDW95rB0ZteQUWXaUELkKFoPn54GK6bIWK3A5PNBRNYMQaiG9mSBuObf2kBZE8wb4dfcqgliHR/gJkMc6G) (Note: In the playground typing for `v-slot` does not seem to work but it works in my editor.)


```vue
<script setup>
import TempVar from './temp-var';
</script>

<template>
  <TempVar :foo="Math.random()" v-slot="{ foo }">
    <div>{{ foo }}</div>
  </TempVar>
</template>
```

```ts
// temp-var.ts
const TempVar = defineComponent(
  (props, ctx) => () => ctx.slots.default?.(ctx.attrs),
  { inheritAttrs: false },
) as unknown as new <T>(props: T) => {
  $props: T;
  $slots: { default?: (context: T) => unknown };
};

export default TempVar;
```

**Comment by unrevised6419** [comment]
I created this library for myself https://www.npmjs.com/package/@allindevelopers/vue-local-scope

The code is pretty simple https://github.com/allindevelopers/vue-local-scope/blob/main/lib/local-scope.component.vue

**LocalScope.vue**

```vue
<script setup lang="ts" generic="T">
defineOptions({ inheritAttrs: false });
defineProps</* @vue-ignore */ T>();
defineSlots<{ default?(props: T): any }>();
</script>

<template>
	<slot v-bind="$attrs as T" />
</template>
```

**Reply by aa-ndrej** [reply]
For me this works perfectly inside of VS Code, but when running `nuxt typecheck` I get type errors like `error TS2339: Property 'x' does not exist on type 'unknown'.` when using it like
```vue
<LocalScope v-slot="{ x }" :x="'y'"> ...
```
Does typechecking work for you, @unrevised6419 ?

I had to change it to the following to make typechecking work for me with `nuxt typecheck`:
```vue
<script setup generic="T" lang="ts">
defineOptions({ inheritAttrs: false })
defineProps<{ slotProps: T }>()
defineSlots<{ default?(props: T): unknown }>()
</script>

<template>
	<slot v-bind="{ ...slotProps }" />
</template>
```

and then use it like
```vue
<LocalScope v-slot="{ x }" :slot-props="{ x: 'foo' }"> ...
```

**Reply by unrevised6419** [reply]
@aa-ndrej will need to check probably something changed in latest Vue version and it needs some adjustments.
Actually my first implementation was similar to your example.

**Reply by unrevised6419** [reply]
Checked. For me this is also reported in vscode.

And it seems this "regression" started with `vue-tsc@3`
Downgrading to `vue-tsc@2` type checking seems to work fine

Still reported because vscode version of `vue-tsc` is not taken from `node_modules`

https://github.com/user-attachments/assets/d293cc21-b735-4115-b8f9-67541cf7a58e

**Reply by unrevised6419** [reply]
Reported an issue https://github.com/vuejs/language-tools/issues/5592

**Reply by unrevised6419** [reply]
The issues, seems to already be fixed, and will be available in the next release.

https://github.com/vuejs/language-tools/pull/5424

**Reply by aa-ndrej** [reply]
Thank you for the quick and excellent follow-up! 🙏🏼

**Comment by michaelKurowski** [comment]
But why do we need workarounds for such a basic feature that should be available in vuejs by default
